packages/backend-for-frontend/src/services/delegationService.ts

```diff
@@ -1,6 +1,187 @@
 
-export function delegationServiceBuilder(_apiClients: PagoPAInteropBeClients) {
-  return {};
+// eslint-disable-next-line max-params
+async function enhanceDelegation<
+  T extends bffApi.Delegation | bffApi.CompactDelegation
+>(
+  delegationClient: DelegationProcessClient,
+  tenantClient: TenantProcessClient,
+  catalogClient: CatalogProcessClient,
+  delegationId: string,
+  headers: Headers,
+  toApiConverter: (
+    delegation: delegationApi.Delegation,
+    delegator: tenantApi.Tenant,
+    delegate: tenantApi.Tenant,
+    eservice: catalogApi.EService
+  ) => T
+): Promise<T> {
+  const delegation: delegationApi.Delegation =
+    await delegationClient.delegation.getDelegation({
+      params: { delegationId },
+    });
+
+  if (!delegation) {
+    throw delegationNotFound(delegationId);
+  }
+
+  const delegator: tenantApi.Tenant = await tenantClient.tenant.getTenant({
+    params: { id: delegation.delegatorId },
+    headers,
+  });
+
+  const delegate: tenantApi.Tenant = await tenantClient.tenant.getTenant({
+    params: { id: delegation.delegateId },
+    headers,
+  });
+
+  const eservice: catalogApi.EService = await catalogClient.getEServiceById({
+    params: { eServiceId: delegation.eserviceId },
+    headers,
+  });
+
+  return toApiConverter(delegation, delegator, delegate, eservice);
+}
+
+export function delegationServiceBuilder(
+  delegationClients: DelegationProcessClient,
+  tenantClient: TenantProcessClient,
+  catalogClient: CatalogProcessClient
+) {
+  return {
+    async getDelegationById(
+      delegationId: DelegationId,
+      { headers, logger }: WithLogger<BffAppContext>
+    ): Promise<bffApi.Delegation> {
+      logger.info(`Retrieving delegation with id ${delegationId}`);
+
+      return enhanceDelegation<bffApi.Delegation>(
+        delegationClients,
+        tenantClient,
+        catalogClient,
+        delegationId,
+        headers,
+        toBffDelegationApiDelegation
+      );
+    },
+    async getDelegations(
+      {
+        limit,
+        offset,
+        states,
+        kind,
+        delegatedIds,
+        delegatorIds,
+      }: {
+        limit: number;
+        offset: number;
+        states?: bffApi.DelegationState[];
+        kind?: bffApi.DelegationKind;
+        delegatedIds?: string[];
+        delegatorIds?: string[];
+      },
+      { headers, logger }: WithLogger<BffAppContext>
+    ): Promise<bffApi.CompactDelegations> {
+      logger.info("Retrieving all delegations");
+
+      const delegationsResults =
+        await delegationClients.delegation.getDelegations({
+          queries: {
+            limit,
+            offset,
+            delegatorIds,
+            delegateIds: delegatedIds,
+            delegationStates: states,
+            kind,
+          },
+        });
+
+      const delegationEnanched = await Promise.all(
+        delegationsResults.results.map((delegation) =>
+          enhanceDelegation<bffApi.CompactDelegation>(
```

Comment:
The frontend will call the listing function in two instances:

- in the `Given delegations` tab, where the `delegator` will always be the same.
- in the `Received delegations` tab, where the `delegate` will always be the same.

I think would be good to implement some sort of caching for the tenants data, se we could avoid make a lot of calls to the process for the same tenant.

What do you think?

Reply:
Handled in this [commit](https://github.com/pagopa/interop-be-monorepo/pull/1152/commits/ed31e66a92b9dd3e74021b80379d0a939a5075b3) of a secondary [PR](https://github.com/pagopa/interop-be-monorepo/pull/1152) related to delegation enhancement.